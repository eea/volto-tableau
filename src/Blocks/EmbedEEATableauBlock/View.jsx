import React from 'react';
import ConnectedTableau from '../../ConnectedTableau/ConnectedTableau';
import { Sources } from '../../Sources';
import { PrivacyProtection } from '@eeacms/volto-embed';

import { StyleWrapperView } from '@eeacms/volto-block-style/StyleWrapper';
import cx from 'classnames';

import { getContent } from '@plone/volto/actions';

import { connect } from 'react-redux';
import { compose } from 'redux';

const View = (props) => {
  const { data_provenance, tableau_visualization } = props || {};
  const data = React.useMemo(() => props.data || {}, [props.data]);
  const { vis_url = '' } = data;
  const show_sources = data?.show_sources ?? false;

  React.useEffect(() => {
    if (vis_url) {
      props.getContent(vis_url, null, props.id);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [vis_url]);

  return (
    <StyleWrapperView
      {...props}
      data={data}
      styleData={{
        ...data.styles,
        customClass: cx(data.styles?.customClass || '', 'embed-container'),
      }}
    >
      <PrivacyProtection data={data} {...props}>
        {data?.vis_url ? (
          <>
            {tableau_visualization?.general?.url ? (
              <>
                <ConnectedTableau
                  {...props.tableau_visualization}
                  id={props.id}
                  {...props}
                />

                {show_sources &&
                data_provenance &&
                data_provenance?.data?.data_provenance &&
                tableau_visualization ? (
                  <Sources sources={data_provenance.data.data_provenance} />
                ) : show_sources ? (
                  <div>Data provenance is not set in the visualization</div>
                ) : (
                  ''
                )}
              </>
            ) : !tableau_visualization?.general?.url ? (
              <div>Url is not set in the visualization</div>
            ) : null}
          </>
        ) : (
          <div>Please select a visualization from block editor.</div>
        )}
      </PrivacyProtection>
    </StyleWrapperView>
  );
};

export default compose(
  connect(
    (state, props) => ({
      data_provenance: state.content.subrequests?.[props.id],
      tableau_visualization:
        state.content.subrequests?.[props.id]?.data?.tableau_visualization_data,
    }),
    {
      getContent,
    },
  ),
)(React.memo(View));
