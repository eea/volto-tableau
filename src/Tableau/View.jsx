import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { toast } from 'react-toastify';
import { Toast } from '@plone/volto/components';
import { setTableauApi } from '@eeacms/volto-tableau/actions';
import cx from 'classnames';
import { loadTableauScript } from '../helpers';

const Tableau = (props) => {
  const ref = React.useRef(null);
  const filters = React.useRef(props.data.filters || {});
  const mounted = React.useRef(false);
  const [viz, setViz] = React.useState(null);
  const {
    canUpdateUrl = true,
    data = {},
    error = null,
    extraFilters = {},
    extraOptions = {},
    loaded = false,
    mode = 'view',
    screen = {},
    setError = () => {},
    setLoaded = () => {},
    version = '2.8.0',
  } = props;
  const {
    autoScale = false,
    hideTabs = false,
    hideToolbar = false,
    sheetname = '',
    toolbarPosition = 'Top',
  } = data;
  const defaultUrl = data.url;
  const url = props.url || defaultUrl;

  //load tableau from script tag
  const tableau = loadTableauScript(() => {}, version);

  const onFilterChange = (filter) => {
    const newFilters = { ...filters.current };
    const fieldName = filter.getFieldName();
    const values = filter
      .getAppliedValues()
      .map((appliedValue) => appliedValue.value);
    newFilters[fieldName] = values;
    if (JSON.stringify(newFilters) !== JSON.stringify(filters)) {
      props.onChangeBlock(props.block, {
        ...data,
        filters: {
          ...newFilters,
        },
      });
      filters.current = { ...newFilters };
    }
  };

  const disposeViz = () => {
    if (viz) {
      viz.dispose();
    }
    if (loaded) {
      setLoaded(false);
    }
    if (error) {
      setError(null);
    }
  };

  const initViz = () => {
    disposeViz();
    try {
      const newViz = new tableau.Viz(ref.current, url || defaultUrl, {
        hideTabs,
        hideToolbar,
        sheetname,
        toolbarPosition,
        ...data.filters,
        ...extraFilters,
        ...extraOptions,
        onFirstInteractive: () => {
          setLoaded(true);
          if (newViz && mode === 'edit') {
            const workbook = newViz.getWorkbook();
            const newData = {
              url: canUpdateUrl ? newViz.getUrl() : defaultUrl,
              sheetname: workbook.getActiveSheet().getName(),
            };
            if (newData.url !== url || newData.sheetname !== sheetname) {
              props.onChangeBlock(props.block, {
                ...data,
                ...newData,
              });
              toast.success(<Toast success title={'Tableau data updated'} />);
            }
            // Filter change event
            newViz.addEventListener(
              tableau.TableauEventName.FILTER_CHANGE,
              (event) => {
                event.getFilterAsync().then((filter) => {
                  onFilterChange(filter);
                });
              },
            );
          }
        },
      });

      return setViz(newViz);
    } catch (e) {
      setError(e._message);
    }
  };

  const addExtraFilters = (extraFilters) => {
    const worksheets = viz.getWorkbook().getActiveSheet().getWorksheets() || [];

    worksheets.forEach((worksheet) => {
      if (worksheet.getSheetType() === tableau.DashboardObjectType.WORKSHEET) {
        Object.keys(extraFilters).forEach((filter) => {
          if (!extraFilters[filter]) {
            worksheet.clearFilterAsync(filter);
          } else {
            worksheet.applyFilterAsync(
              filter,
              extraFilters[filter],
              tableau.FilterUpdateType.REPLACE,
            );
          }
        });
      }
    });
  };

  const updateScale = () => {
    const tableauWrapper = ref.current;
    const tableau = tableauWrapper?.querySelector('iframe');
    if (mounted.current && viz && tableauWrapper && tableau) {
      const { sheetSize = {} } = viz.getVizSize() || {};
      const vizWidth = sheetSize?.minSize?.width || 1;
      const vizHeight = sheetSize?.minSize?.height || 0;
      const scale = Math.min(tableauWrapper.clientWidth / vizWidth, 1);
      tableau.style.transform = `scale(${scale})`;
      tableau.style.width = `${100 / scale}%`;
      tableauWrapper.style.height = `${scale * vizHeight}px`;
    }
  };

  React.useEffect(() => {
    if (!mounted.current) {
      mounted.current = true;
    }
    return () => {
      mounted.current = false;
    };
  }, []);

  React.useEffect(() => {
    // Load new tableau api
    if (__CLIENT__ && !props.tableau[version]) {
      props.setTableauApi(version, props.mode);
    }
    if (__CLIENT__) {
      loadTableauScript(() => {}, version);
    }
    /* eslint-disable-next-line */
  }, [version]);

  React.useEffect(() => {
    if (__CLIENT__ && tableau && url) {
      initViz();
    } else {
      disposeViz();
    }

    return () => {
      disposeViz();
    };
    /* eslint-disable-next-line */
  }, [
    hideTabs,
    hideToolbar,
    autoScale,
    sheetname,
    tableau,
    toolbarPosition,
    url,
    version,
  ]);

  React.useEffect(() => {
    if (mounted.current && loaded && viz) {
      addExtraFilters(extraFilters);
    }
    /* eslint-disable-next-line */
  }, [JSON.stringify(extraFilters)]);

  React.useEffect(() => {
    if (autoScale) {
      updateScale();
    }
    /* eslint-disable-next-line */
  }, [loaded, screen?.page?.width]);

  return (
    <div id="tableau-wrap">
      <div id="tableau-outer">
        {data && Object.keys(data).length > 0 ? (
          <>
            {loaded ? (
              ''
            ) : (
              <div className="tableau-loader">
                <span>
                  {mode === 'edit'
                    ? 'Loading...'
                    : `Loading Tableau v${version}`}
                </span>
              </div>
            )}
          </>
        ) : (
          <div>No data present in that visualization.</div>
        )}
        <div
          className={cx('tableau', version, {
            'tableau-scale': autoScale,
          })}
          ref={ref}
        />
      </div>
    </div>
  );
};

export default compose(
  connect(
    (state, props) => ({
      tableau: state.tableau,
      screen: state.screen,
    }),
    { setTableauApi },
  ),
)(Tableau);
