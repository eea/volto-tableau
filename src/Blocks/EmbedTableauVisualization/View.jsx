import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { flattenToAppURL } from '@plone/volto/helpers';
import { getContent } from '@plone/volto/actions';
import { PrivacyProtection } from '@eeacms/volto-embed';
import Tableau from '@eeacms/volto-tableau/Tableau/Tableau';

const View = (props) => {
  const data = props.data;
  const {
    with_note = true,
    with_sources = true,
    with_more_info = true,
    with_download = true,
    with_share = true,
    tableau_height = '700',
  } = data;
  const { figure_note = [], data_provenance = {}, tableau_visualization } =
    props.tableau_visualization_data || {};

  const tableau_vis_url = flattenToAppURL(data.tableau_vis_url || '');

  React.useEffect(() => {
    if (tableau_vis_url) {
      props.getContent(tableau_vis_url, null, props.id);
    }
    // eslint-disable-next-line
  }, [tableau_vis_url]);

  return (
    <div className="embed-tableau">
      <PrivacyProtection
        {...props}
        data={{ ...data, url: tableau_visualization?.url }}
      >
        {!tableau_vis_url && (
          <div>Please select a tableau visualization from block editor.</div>
        )}
        {!!tableau_vis_url && (
          <>
            {!tableau_visualization?.url && props.mode === 'edit' && (
              <div>Url is not set in the visualization</div>
            )}
            {!!tableau_visualization?.url && (
              <Tableau
                data={{
                  ...tableau_visualization,
                  with_note,
                  with_sources,
                  with_more_info,
                  with_download,
                  with_share,
                  tableau_height,
                  tableau_vis_url,
                }}
                figure_note={figure_note}
                sources={data_provenance.data || []}
              />
            )}
          </>
        )}
      </PrivacyProtection>
    </div>
  );
};

export default compose(
  connect(
    (state, props) => ({
      tableau_visualization_data: state.content.subrequests?.[props.id]?.data,
    }),
    {
      getContent,
    },
  ),
)(React.memo(View));
