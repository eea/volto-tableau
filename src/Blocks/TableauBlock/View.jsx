import React from 'react';
import { connect } from 'react-redux';
import Tableau from '@eeacms/volto-tableau/Tableau/Tableau';
import qs from 'query-string';
import config from '@plone/volto/registry';
import '@eeacms/volto-tableau/less/tableau.less';

const View = React.forwardRef((props, ref) => {
  const viz = React.useRef();
  const [vizState, setVizState] = React.useState({
    loaded: false,
    loading: false,
    error: null,
  });
  const [extraFilters, setExtraFilters] = React.useState({});
  const { block, data = {}, query = {}, onChangeBlock, mode } = props;
  const {
    urlParameters = [],
    staticParameters = [],
    title = null,
    description = null,
    with_sources = true,
    with_download = true,
    with_share = true,
    with_enlarge = true,
    sources,
  } = data;

  const extraOptions = React.useMemo(() => {
    const options = {};
    staticParameters.forEach((parameter) => {
      if (parameter.field && parameter.value) {
        options[parameter.field] = parameter.value;
      }
    });
    return options;
  }, [staticParameters]);

  React.useEffect(() => {
    const newFilters = { ...extraFilters };
    urlParameters.forEach((element) => {
      if (element.field && typeof query[element.urlParam] !== 'undefined') {
        newFilters[element.field] = query[element.urlParam];
      } else if (newFilters[element.field]) {
        delete newFilters[element.field];
      }
    });
    setExtraFilters(newFilters);
    /* eslint-disable-next-line */
  }, [JSON.stringify(query), JSON.stringify(urlParameters)]);

  React.useImperativeHandle(ref, () => {
    return viz.current;
  });

  React.useEffect(() => {
    if (props.setVizState) {
      props.setVizState(vizState);
    }
    /* eslint-disable-next-line */
  }, [vizState]);

  return (
    <div className="tableau-block">
      {vizState.loaded && title ? (
        <h3 className="tableau-title">{title}</h3>
      ) : (
        ''
      )}
      {vizState.loaded && description ? (
        <p className="tableau-description">{description}</p>
      ) : (
        ''
      )}
      <Tableau
        ref={viz}
        mode={mode}
        block={block}
        data={{
          ...data,
          with_sources,
          with_download,
          with_share,
          with_enlarge,
        }}
        sources={sources}
        extraFilters={extraFilters}
        extraOptions={extraOptions}
        breakpoints={config.blocks.blocksConfig.tableau_block.breakpoints}
        setVizState={setVizState}
        onChangeBlock={onChangeBlock}
      />
    </div>
  );
});

export default connect(
  (state) => ({
    query: {
      ...(qs.parse(state.router.location?.search?.replace('?', '')) || {}),
      ...(state.discodata_query?.search || {}),
    },
    screen: state.screen,
  }),
  null,
  null,
  { forwardRef: true },
)(View);
