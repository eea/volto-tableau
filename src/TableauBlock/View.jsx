import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { withRouter } from 'react-router';
import Tableau from '@eeacms/volto-tableau/Tableau/View';
import config from '@plone/volto/registry';
import { getLatestTableauVersion } from '@eeacms/volto-tableau/Tableau/tableau-api';
import qs from 'querystring';
import './styles.less';

const getDevice = (config, width) => {
  const breakpoints = config.blocks.blocksConfig.tableau_block.breakpoints;
  let device = 'default';
  Object.keys(breakpoints).forEach((breakpoint) => {
    if (
      width <= breakpoints[breakpoint][0] &&
      width >= breakpoints[breakpoint][1]
    ) {
      device = breakpoint;
    }
  });
  return device;
};

const View = (props) => {
  const [extraFilters, setExtraFilters] = React.useState({});
  const { data = {}, query = {}, screen = {} } = props;
  const {
    breakpointUrls = [],
    urlParameters = [],
    title = null,
    description = null,
  } = data;
  const version =
    props.data.version ||
    config.settings.tableauVersion ||
    getLatestTableauVersion();
  const device = getDevice(config, screen.screenWidth || Infinity);
  const breakpointUrl = breakpointUrls.filter(
    (breakpoint) => breakpoint.device === device,
  )[0]?.url;
  const url = breakpointUrl || data.url;

  React.useEffect(() => {
    const newExtraFilters = { ...extraFilters };
    urlParameters.forEach((element) => {
      if (element.field && typeof query[element.urlParam] !== 'undefined') {
        newExtraFilters[element.field] = query[element.urlParam];
      } else if (newExtraFilters[element.field]) {
        delete newExtraFilters[element.field];
      }
    });
    setExtraFilters(newExtraFilters);
    /* eslint-disable-next-line */
  }, [JSON.stringify(query), JSON.stringify(urlParameters)])

  return (
    <div className="tableau-block">
      {props.mode === 'edit' ? (
        <>
          <h3>Tableau {version}</h3>
          {!props.data.url ? <p>URL required</p> : ''}
        </>
      ) : (
        ''
      )}
      {title ? <h3>{title}</h3> : ''}
      {description ? <p>{description}</p> : ''}
      <Tableau
        {...props}
        url={url}
        canUpdateUrl={!breakpointUrl}
        version={version}
        extraFilters={extraFilters}
        extraOptions={{ device }}
      />
    </div>
  );
};

export default compose(
  connect((state, props) => ({
    query: {
      ...(qs.parse(state.router.location.search.replace('?', '')) || {}),
      ...(state.discodata_query.search || {}),
    },
    screen: state.screen,
  })),
)(withRouter(View));
