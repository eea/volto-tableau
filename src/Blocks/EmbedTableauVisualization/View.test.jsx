import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import View from './View';

const mockStore = configureStore([]);
const store = mockStore({ content: { subrequests: [] } });

window.URL.createObjectURL = jest.fn(() => 'test');

jest.mock('@eeacms/volto-embed', () => ({
  PrivacyProtection: ({ children }) => <div>{children}</div>,
}));

jest.mock('@plone/volto/components', () => ({
  Icon: ({ children }) => <img alt="incon">{children}</img>,
}));

jest.mock('@plone/volto/components', () => ({
  Toast: ({ children }) => <p>{children}</p>,
}));

describe('View', () => {
  const data = {
    '@type': 'embed_tableau_visualization',
    dataprotection: {},
    tableau_vis_url: 'http://localhost:3000/tableau-ct',
    with_download: true,
    with_more_info: true,
    with_note: true,
    with_share: true,
  };

  it('should render the component', () => {
    const { container } = render(
      <Provider store={store}>
        <View data={data} />
      </Provider>,
    );
    expect(container.querySelector('.embed-container')).toBeInTheDocument();
  });
});
