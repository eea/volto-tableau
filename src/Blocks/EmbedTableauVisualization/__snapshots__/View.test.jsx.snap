// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View should render the component 1`] = `
<div
  className="embed-tableau"
>
  <div
    className="privacy-protection-wrapper"
    style={
      Object {
        "height": "auto",
        "minHeight": "200px",
        "overflow": "hidden",
        "position": "relative",
      }
    }
  >
    <div
      className="tableau-wrapper"
    >
      <div
        className="tableau tableau-2.8.0"
        style={Object {}}
      />
    </div>
  </div>
</div>
`;
